### DISCUSSION TITLE
Flash with static/predefined region/channels/etc what is the std way?

### DISCUSSION THREAD

**Discussion by mariotti** [meshtastic/Discussions#72] [General]
Hi all,

So given the title question: how can I flash with presets?

I see the the files:

[userPrefs.json](https://github.com/meshtastic/firmware/blob/master/userPrefs.json)

and

[userPrefs.h](https://github.com/meshtastic/firmware/blob/master/userPrefs.h)

But they look like more as a demo and not an actual file loaded while flashing.
The values are actually not loaded if I flash directly from the checkedout branch.

Is there a way to achieve this?

Of course the requirement would be that any serial client cannot overwrite
the presets.

Btw, I start to have a lot of fun with meshtastic! Thanks!
A+
F

**Comment by fifieldt** [comment]
These files are baked in to the firmware when it is built. Are you comfortable building your own firmware?

**Reply by mariotti** [reply]
Thanks!

Yes.

I actually just noticed that when building with vscode, as from instructions from the main README, the userPrefs.h is read, and few cpp files are recompiled.
[The userPrefs.json seems to be ignored at present. Though seems to be used in GA]

But via vscode, pushing the FW seems to keep the device original channels configuration.
How can I add flags to the build command? and which flag would make it erase/flash?

To be honest I am more used to command line for building, se that would work too :) 

Right now I am trying to find a way to also do not allow STD clients to not modify the settings.

Thanks Again
any help appreciated

But as I go on I will post here my findings :)

**Reply by mariotti** [reply]
Ok! One option I will try is to flash via the web interface with the erase before flashing option, just to test I am right. But would not be the best process.

**Reply by fifieldt** [reply]
Right, so the userprefs.h won't override settings that have been explicitly specified by the user. Think of it more of something an "OEM" would use.

To attain a device where users can't modify settings, you'd need to alter nodedb.cpp

**Reply by mariotti** [reply]
As I see it I also would need the Channels.cpp

But actually both will not grant to override the existing device settings. Nor that a client cannot change the settings.

But let me experiment a bit. I will be back here.

I have the additional issue that flashing takes time and my t-echo seems a bit odd with flashing,
same process, works depending on some other random parameters :(

**Reply by mariotti** [reply]
Some though.

I had a look at the code and it looks like that while the userPrefs.h is loaded you still need a clean/flash process.
Which I cannot do right now on the t-echo because it has an odd flashing process. I will try on a different device.

On the other side using the userPrefs.h works to set the initial values but if you want to make them static at
present you would need to work with the AdminModule.cpp which apparently is the only one applying changes.

Here I still need to find the best way, which probably is using the predefined process.

PS:
On the t-echo I tried also the command line "meshtastic" with: ```meshtastic --factory-reset-device```
but the result was to hop in a previous configuration. Whatever I had there before.

**Reply by mariotti** [reply]
I found a path to follow.

 - Write your own userPrefs.h
 - build
 - flash with "erase" if possible
 - set up the remote admin 

What I do not know is how to send admin messages, yet :( 

But this should give:
 - You have a preconfigured device
 - Only admin can change settings

This should be grant by this line:

https://github.com/meshtastic/firmware/blob/master/src/modules/AdminModule.cpp#L79

i.e. you can have a managed device.

The part I am missing is how to send admin messages.

cheers
F

**Comment by mariotti** [comment]
In this discussion there are 2 topics:

1 - flash with predefs
2 - grant limited access to settings

I think that flash with predefs is solved, as it also depend on the device.

For grant limited access to settings we have different issues: admin and UI.
Also the intrinsic admin module.

So I close this discussion and open a new one a bit more dedicated.

The idea is to grant access to settings only via hardware access.

See you in the next topic.

**Comment by mariotti** [comment]
Ahh yes I close with the previous comment.
